cpp/src/io/orc/writer_impl.cu

```diff
@@ -2133,15 +2152,55 @@ stripe_dictionaries build_dictionaries(orc_table_view& orc_table,
   gpu::collect_map_entries(stripe_dicts, stream);
   gpu::get_dictionary_indices(stripe_dicts, orc_table.d_columns, stream);
 
-  // Clear map slots; hash map storage is deallocated at the end of this function
-  auto device_dicts_flat = stripe_dicts.device_view().flat_view();
-  thrust::for_each(rmm::exec_policy(stream),
-                   device_dicts_flat.begin(),
-                   device_dicts_flat.end(),
-                   [] __device__(auto& sd) { sd.map_slots = {}; });
-  stripe_dicts.device_to_host_async(stream);
+  // deallocate hash map storage, unused after this point
+  hash_maps_storage.clear();
+
+  // Clear map slots and attach order buffers
+  auto dictionaries_flat = stripe_dicts.host_view().flat_view();
+  for (auto& sd : dictionaries_flat) {
+    if (not sd.is_enabled) { continue; }
+
+    sd.map_slots = {};
+    dict_order_owner.emplace_back(sd.entry_count, stream);
+    sd.data_order = dict_order_owner.back();
+  }
+  stripe_dicts.host_to_device_async(stream);
+
+  // Sort stripe dictionaries alphabetically
+  auto streams = cudf::detail::fork_streams(stream, std::min<size_t>(dict_order_owner.size(), 8));
+  auto stream_idx = 0;
+  for (auto& sd : dictionaries_flat) {
+    if (not sd.is_enabled) { continue; }
+
+    auto const& current_stream = streams[stream_idx];
+
+    // Sort the dictionary data and create a mapping from the sorted order to the original
+    thrust::sequence(
+      rmm::exec_policy_nosync(current_stream), sd.data_order.begin(), sd.data_order.end());
+    thrust::sort_by_key(rmm::exec_policy_nosync(current_stream),
+                        sd.data.begin(),
+                        sd.data.end(),
+                        sd.data_order.begin(),
+                        string_rows_less{orc_table.d_columns, sd.column_idx});
+
+    // Create the inverse permutation - i.e. the mapping from the original order to the sorted
+    auto order_copy = cudf::detail::make_device_uvector_async<uint32_t>(
+      sd.data_order, current_stream, rmm::mr::get_current_device_resource());
+    thrust::scatter(rmm::exec_policy_nosync(current_stream),
+                    thrust::make_counting_iterator<uint32_t>(0),
+                    thrust::make_counting_iterator<uint32_t>(sd.data_order.size()),
```

Comment:
```suggestion
                    thrust::counting_iterator<uint32_t>(0),
                    thrust::counting_iterator<uint32_t>(sd.data_order.size()),
```
I've been told that if you specify the type, there is no reason to use the `make_counting_iterator` to make a `counting_iterator`.